internal/cimfs/cim_reader_windows.go

```diff
@@ -0,0 +1,1070 @@
+			return 0, err
+		}
+		reg[i].f = rf
+		fi, err := rf.Stat()
+		if err != nil {
+			return 0, err
+		}
+		reg[i].size = uint64(fi.Size())
+		var rh format.RegionHeader
+		err = readBin(rf, &rh)
+		if err != nil {
+			return 0, fmt.Errorf("reading region header %s: %s", name, err)
+		}
+		err = validateHeader(&rh.Common)
+		if err != nil {
+			return 0, fmt.Errorf("validating region header %s: %s", name, err)
+		}
+	}
+	return int(rs.Count), nil
+}
+
+func readFilesystemHeader(f *os.File) (format.FilesystemHeader, error) {
+	var fsh format.FilesystemHeader
+	err := readBin(f, &fsh)
+	if err != nil {
+		return fsh, fmt.Errorf("reading filesystem header: %s", err)
+	}
+	err = validateHeader(&fsh.Common)
+	if err != nil {
+		return fsh, fmt.Errorf("validating filesystem header: %s", err)
+	}
+	return fsh, nil
+}
+
+// Open opens a CIM file for read access.
+func Open(p string) (_ *CimFsReader, err error) {
+	defer func() {
+		if err != nil {
+			err = &OpError{Cim: p, Op: "open", Err: err}
+		}
+	}()
+	f, err := os.Open(p)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	fsh, err := readFilesystemHeader(f)
+	if err != nil {
+		return nil, err
+	}
+	parents := make([]format.RegionSet, fsh.ParentCount)
+	err = readBin(f, parents)
+	if err != nil {
+		return nil, fmt.Errorf("reading parent region sets: %s", err)
+	}
+	regionCount := int(fsh.Regions.Count)
+	for i := range parents {
+		regionCount += int(parents[i].Count)
+	}
+	if regionCount == 0 || regionCount > 0x10000 {
+		return nil, fmt.Errorf("invalid region count %d", regionCount)
+	}
+	c := &CimFsReader{
+		name:       p,
+		reg:        make([]region, regionCount),
+		upcase:     make([]uint16, format.UpcaseTableLength),
+		inodeCache: make(map[format.FileID]*inode),
+		sdCache:    make(map[format.RegionOffset][]byte),
+	}
+	defer func() {
+		if err != nil {
+			c.Close()
+		}
+	}()
+
+	imagePath := filepath.Dir(p)
+	reg := c.reg
+	for i := range parents {
+		n, err := loadRegionSet(&parents[i], imagePath, reg)
+		if err != nil {
+			return nil, err
+		}
+		reg = reg[n:]
+	}
+	_, err = loadRegionSet(&fsh.Regions, imagePath, reg)
+	if err != nil {
+		return nil, err
+	}
+	var fs format.Filesystem
+	err = c.readBin(&fs, fsh.FilesystemOffset, 0)
+	if err != nil {
+		return nil, fmt.Errorf("reading filesystem info: %s", err)
+	}
+	c.ftables = make([]fileTable, fs.FileTableDirectoryLength)
+	c.ftdes = make([]format.FileTableDirectoryEntry, fs.FileTableDirectoryLength)
+	err = c.readBin(c.ftdes, fs.FileTableDirectoryOffset, 0)
+	if err != nil {
+		return nil, fmt.Errorf("reading file table directory: %s", err)
+	}
+	err = c.readBin(c.upcase, fs.UpcaseTableOffset, 0)
+	if err != nil {
+		return nil, fmt.Errorf("reading upcase table: %s", err)
+	}
+	c.root, err = c.getInode(fs.RootDirectory)
+	if err != nil {
+		return nil, fmt.Errorf("reading root directory: %s", err)
+	}
+	return c, nil
+}
+
+// Close releases resources associated with the Cim.
+func (cr *CimFsReader) Close() error {
```

Comment:
Looks like this always returns nil, is this just to satisfy the `io.Closer` interface?

Reply:
Hmm, this is mostly John's code so I am not sure if he had this in mind but I think it's good that this satisfies the `io.Closer` interface and hence also the `ReaderCloser` interface. So let's keep this.

Reply:
No I agree it should, I was just wondering if that was the reason